File: permutations_string/permutations.cpp
// Find all the permuations of a string. 

#include <iostream>
#include <string>
using namespace std;

void swapChar (char &a, char &b) 
{
	char temp; 
	temp = a;
  a = b;
  b = temp;
}

void permutate (string strPermutations, int start, int end) 
{
  static int count;
  
  if ( start == end ) 
  {
    cout << ++count << " : " << strPermutations << endl;
    return; //- Return is here for clarity.
  } 
  else 
  {
    for ( int i = start; i < strPermutations.length(); ++i ) 
    {
      swapChar(strPermutations[start], strPermutations[i]);
      permutate(strPermutations,start+1, end);
      swapChar(strPermutations[start], strPermutations[i]);
    }
  }
}

int main() 
{
	int    pause;
	string permuteStr;

	cout << "Enter the string : " << endl;
	
  cin >> permuteStr;
	
  cout << "The permutations of the given string : " << endl;
	
  permute(permuteStr, 0, permuteStr.length());
	
  cout << "Enter in a digit and press Enter to unpause." << endl;
	
  cin >> pause;
	
  return 0;
}

File: permutations_string/permutations.js
// Find all permutations of a string. 

function setCharAt(str,index,chr) {
    if(index > str.length-1) return str;
    return str.substr(0,index) + chr + str.substr(index+1);
}

function swapChars (index1,index2, strToSwap) {
  var tempChar = strToSwap[index1];
  return setCharAt( setCharAt(strToSwap, index1, strToSwap[index2]), index2, tempChar );
}

function permutate(strPermutations, start, end) {
  if ( start === end ) {
    console.log( strPermutations );
    return; //- Return is here for extra clarity
  } else {
    for ( var i = start; i < strPermutations.length; ++i ) {
      strPermutations = swapChars( start, i, strPermutations);
      permutate( strPermutations, start + 1, end );
      strPermutations = swapChars( start, i, strPermutations);
    }
  }
}

// Testing the permutate function
permutate("abc",0,3);
permutate("abcd",0,4);

